stonesoup/sensor/actionable.py

```diff
@@ -0,0 +1,179 @@
+# -*- coding: utf-8 -*-
+import datetime
+import inspect
+from abc import ABC
+from abc import abstractmethod
+from typing import Set, Sequence
+
+from .action import Action, ActionGenerator
+from ..base import Base, Property
+
+
+class ActionableProperty(Property):
+    """Property that is modified via an :class:`~.Action` with defined, non-equal start and end
+    times."""
+
+    def __init__(self, generator_cls,
+                 cls=None, *, default=inspect.Parameter.empty,
+                 doc=None, readonly=False):
+        super().__init__(cls=cls, default=default, doc=doc, readonly=readonly)
+        self.generator_cls = generator_cls
+
+
+class Actionable(Base, ABC):
+    """Base Actionable type.
+
+    Contains the core methods of an actionable sensor/platform type.
+
+    Notes
+    -----
+    An Actionable is required to have a `timestamp` attribute, in order to validate actions and
+    act. This is an abstract base class, and not intended for direct use. Attaining a timestamp is
+    left to the inheriting type.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.scheduled_actions = dict()  # dictionary of property - action pairs
+
+    @abstractmethod
+    def validate_timestamp(self):
+        """Method of attaining actionable's timestamp.
+
+        Returns
+        -------
+        bool
+            True if timestamp has been successfully set, False otherwise.
+        """
+        raise NotImplementedError()
+
+    @property
+    def _actionable_properties(self):
+        """Dictionary of all name - property pairs where the property is an
+        :class:`~.ActionableProperty` (i.e. it is modified via action)."""
+        return {_name: _property for _name, _property in type(self).properties.items()
+                if isinstance(_property, ActionableProperty)}
+
+    def _default_action(self, name, property_, timestamp):
+        """Returns the default action of the action generator associated with the property
+        (assumes the property is an :class:`~.ActionableProperty`."""
+        generator = property_.generator_cls(owner=self,
+                                            attribute=name,
+                                            start_time=self.timestamp,
+                                            end_time=timestamp)
+        return generator.default_action
+
+    def actions(self, timestamp: datetime.datetime, start_timestamp: datetime.datetime = None
```

Comment:
The action-generators that this method creates will determine their corresponding actionable attribute initial value with `getattr(self.owner, self.attribute)`. That is the value of the attribute at the sensor's "current" timestamp, and not necessarily at the `start_timestamp` that is passed to this method. So this `start_timestamp` argument is irrelevant unless we change logic in the action generators so that they calculate what the attribute value WILL be at `start_timestamp` and determine possible actions from there.
I can't recall why we required this argument in the first place.